Run doctests in pytest. Add linters to pre-commit. Add a pre-commit workflow.

aliddell opened the pull request:
- Adds doctests for pyi and md files.
- Pulls pytest arguments used in build / test workflows into pyproject.toml.
- Adds [ruff](https://docs.astral.sh/ruff/) and [rustfmt](https://github.com/rust-lang/rustfmt) checks to pre-commit.
- Adds a pre-commit check workflow that runs on PRs into main.

aliddell commented on README.md:
````diff
@@ -35,31 +35,30 @@ The provided [napari][] plugin is a good example of how to stream for visualizat
-runtime.start()
-runtime.stop()  # wait for acquisition to complete
+```python-repl
+>>> import acquire
+>>> runtime = acquire.Runtime()
+>>> dm = runtime.device_manager()
+>>> props = runtime.get_configuration()
+>>> # select the first Hamamatsu camera
+>>> props.video[0].camera.identifier = dm.select(acquire.DeviceKind.Camera, "hamamatsu.*")
+>>> # stream to zarr
+>>> props.video[0].storage.identifier = dm.select(acquire.DeviceKind.Storage, "zarr")
+>>> props.video[0].storage.settings.filename = "out.zarr"
+>>> props.video[0].camera.settings.shape = (2304, 2304)
+>>> props.video[0].camera.settings.pixel_type = acquire.SampleType.U16
+>>> props.video[0].max_frame_count = 100
+>>> props = runtime.set_configuration(props)
+>>> runtime.start()  # doctest: +SKIP
+>>> # wait for acquisition to complete
+>>> runtime.stop()  # doctest: +SKIP
````
@andy-sweet I kind of hate that doctest is forcing this to look like a REPL session. I'd appreciate any suggestions.

andy-sweet replied:
While I like the `>>>` in Python source because it helps distinguish code from docs, agreed that I don't like it here in Markdown files.

I guess there are some ways to omit the `>>>` and extract the source as something to test and I even found [a pytest plugin that does this](https://github.com/modal-labs/pytest-markdown-docs).

But I'd lean towards just leaving the `>>>` in because certain Python repls (e.g. ipython) will just ignore them anyway.

nclack reviewed (approved):
Looks good to me.

I know you've got a question for @andy-sweet and I wonder about other things to get into the pre commit (e.g. `ruff` and `cargo fmt`). So feel free to hold off on merging till you're comfortable. But I think this can go in as is.

andy-sweet reviewed (changes requested):
Running doctests as part of pre-commit feels a bit heavy to me, especially when those tests depend on locally connected hardware.

I think that [pytest should just pick up on any doctests in Python modules by default](https://docs.pytest.org/en/7.1.x/how-to/doctest.html) and we can modify the glob pattern to also look at the Markdown files (and Rust files if needed).

andy-sweet commented on tests/test_docs.py:
```diff
@@ -0,0 +1,34 @@
+
+
+def test_rust_sources(base_dir):
+    for f in base_dir.glob("src/**/*.rs"):
```
Can you give an example of a Python doctest in the Rust source code? Or what you might expect that to look like?

aliddell replied:
You're right, there actually aren't any.

aliddell commented:
I added a check on the .pyi files to test_docs.py and removed the pre-commit check, opting to just call `pytest -k test_docs` from the usual tests workflow.

andy-sweet commented on tests/test_docs.py:
```diff
@@ -0,0 +1,31 @@
+import doctest
```
If we use the `--doctest-glob` options as in the github workflows and also use `--doctest-modules`, I don't think we need this explicit file to collect doctests. By including it, we might actually run duplicates, though maybe pytest ignores those?

andy-sweet commented on .github/workflows/test_pr.yml:
```diff
@@ -144,8 +145,7 @@ jobs:
         run: |
-          python -m pytest -k test_egrabber --color=yes --cov-report=xml --cov=acquire --maxfail=5 --log-cli-level=0
-
+          python -m pytest -k test_egrabber --color=yes --cov-report=xml --cov=acquire --maxfail=5 --log-cli-level=0 --doctest-glob="*.pyi" --doctest-glob="*.md"
```
Instead of repeatedly specifying `--doctest-glob`, it's probably better to put these in the pytest.ini options `addopts` in the `pyproject.toml`. Unless there's a reason to only run the doctests for some jobs?

andy-sweet commented on .github/workflows/test_pr.yml:
```diff
@@ -53,6 +53,7 @@ jobs:
       - name: Test
         run: |
           python -m pytest -k test_basic --color=yes --cov-report=xml --cov=acquire --maxfail=5 --log-cli-level=0
+          python -m pytest -k test_docs --color=yes --cov-report=xml --cov=acquire --maxfail=5 --log-cli-level=0
```
Is there a difference between explicitly collecting doctests with `test_docs` and using the pytest doctest options?

If not, I think we should make these consistent (ideally through the pytest options in the pyproject.toml file). See other comments for more context.

andy-sweet commented on pyproject.toml:
```diff
@@ -61,3 +62,13 @@ addopts = "--tb=short -s"
+
+[tool.ruff]
+exclude = [
+    "*.pyi"
```
Is adding ruff directly related to this PR? If not, that's OK, but why ignore the pyi file?

andy-sweet reviewed (changes requested):
Mostly looks good, but I don't think we need to explicitly collect doctests with `test_docs.py`, in which case we can and should make a few simplifications.

andy-sweet commented on pyproject.toml:
```diff
@@ -57,7 +58,17 @@ profile = "black"
 log_cli = true # when true, messages are printed immediately
+
+[tool.ruff]
+exclude = [
+    "__init__.py"
+]
+
+ignore = [
+    "F403",
+    "F405",
+]
```
I'm OK to start with these excludes and ignores, though I'd prefer to have none of them in the long term.

andy-sweet reviewed (approved):
This PR is mostly about adding code format checkers to pre-commit (and the resulting changes that causes) than it is about running the tests on documented examples.

I'm fine with that (i.e. no need to split the PR), but at least update the title and description to match that.
